pkg/controller/hosts_reconciler.go

```diff
@@ -0,0 +1,230 @@
+// SPDX-FileCopyrightText: 2023-present Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package controller
+
+import (
+	"context"
+	"github.com/onosproject/onos-api/go/onos/topo"
+	"github.com/onosproject/topo-discovery/pkg/southbound"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// HostReconciler provides state and context required for host discovery and reconciliation
+type HostReconciler struct {
+	southbound.HostListener
+	hostDiscovery southbound.HostDiscovery
+	topoClient    topo.TopoClient
+	ctx           context.Context
+	lock          sync.RWMutex
+
+	// Map of agent-id to topo object required to resolve link reports to a device
+	agentDevices map[string]*topo.Object
+}
+
+// NewHostReconciler creates a new host reconciler context
+func NewHostReconciler(ctx context.Context, topoClient topo.TopoClient) *HostReconciler {
+	return &HostReconciler{
+		topoClient:    topoClient,
+		ctx:           ctx,
+		agentDevices:  make(map[string]*topo.Object),
+		hostDiscovery: southbound.NewGNMIHostDiscovery(),
+	}
+}
+
+// DiscoverHosts discovers hosts and reconciles their topology entity counterparts
+func (r *HostReconciler) DiscoverHosts(object *topo.Object) {
+	// Connect to the host agent gNMI server and get its agent ID and a map of hosts
+	hostReport, err := r.hostDiscovery.GetHosts(object, r)
+	if err != nil {
+		log.Warnf("Unable to get hosts from device host agent %s: %+v", object.ID, err)
+		return
+	}
+
+	// Register the report and agent ID
+	hostsToProcess := r.registerReport(object, hostReport)
```

Comment:
Here, let's just iterate over the hosts in the report directly.

Reply:
Done